nmsg/kafkaio.c

```
@@ -587,31 +616,57 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
 	my_free(ctx);
 }
 
-static void
-_kafka_error_cb(rd_kafka_t *rk, int err, const char *reason, void *opaque)
-{
-	kafka_ctx_t ctx = (kafka_ctx_t) opaque;
-	rd_kafka_resp_err_t err_kafka = (rd_kafka_resp_err_t) err;
-	if (ctx == NULL) {
-		_nmsg_dprintf(2, "%s: unexpected Kafka opaque is NULL", __func__);
-		return;
-	}
-	switch (err_kafka) {
+static int
+_kafka_handle_error(rd_kafka_t *rk, kafka_ctx_t ctx, rd_kafka_resp_err_t err) {
+	int level = 2;
+
+	switch (err) {
 		/* Keep retrying on socket disconnect, brokers down and message timeout */
 		case RD_KAFKA_RESP_ERR__TRANSPORT:
 		case RD_KAFKA_RESP_ERR__ALL_BROKERS_DOWN:
+			if (ctx->connected)
+				_kafka_set_connected(ctx, __func__, false);
+			break;
 		case RD_KAFKA_RESP_ERR__MSG_TIMED_OUT:
-			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
+			ctx->timedout++;
+			ctx->timedout_total++;
+			if (ctx->timedout > 0 && (ctx->timedout % 100) == 0)
+				_nmsg_dprintf(2,"%s: %ld messages timed out\n", __func__, ctx->timedout);
+			level = 5;
 			break;
+		/* Let librdkafka handle recovery from these errors and other */
 		case RD_KAFKA_RESP_ERR__UNKNOWN_PARTITION:
 		case RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART:
 		case RD_KAFKA_RESP_ERR_OFFSET_OUT_OF_RANGE:
-		/* At the moment treat any broker's error as fatal */
 		default:
-			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
-			_kafka_set_state(ctx, __func__, kafka_state_break);
 			break;
 	}
+
+	if (ctx->last_error != err) {
+		if (ctx->last_error_count > 1)
+			_nmsg_dprintf(level, "%s: Suppressed %ld errors %d\n", __func__, ctx->last_error_count, ctx->last_error);
```

I think `last_error_count` errors were received but `last_error_count - 1` errors were suppressed

---

Fixed